### PR TITLE
[WIP] add World Conquest II to mainline

### PR DESCRIPTION
**Pull request by gfgtdf**
See Also https://github.com/wesnoth/wesnoth/issues/4895 for the issue, 

This is not done yet things to do:

1. Check which terrains we really want in mainline and remove the yz letters in them. DONE
2. Maybe fix https://github.com/wesnoth/wesnoth/issues/4894 LATER
3. Make sure it translates well. (in particular make sure there is no strings appear in hosts language bug) DONE
4. Maybe mainline the "step on item" event. LATER?
5. Run optiong over the images before merging  them. DONE
6. It probably has some unused strings in the cfg files FIXED

There is also some space left for improvement for things that shodul better be done after its merged:

1. Improve the random names generator
 
What i am currently wondering is: this addon used the 1.10 tropical forest terrain because the original author says it looks better near wamt or just for soem variation, but it feels wrong to readd a deleted terrain back to wesnoth, any opinion on how to handle this?

### CODE REVIEW

File: data/campaigns/World_Conquest_II/MODDING.txt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+```
+[event]
+	name = "wc2_init_enemy"
+	[filter_conditional]
+		[variable]
+			name="wc2_enemy_army.length"
+			equals=0
+		[/variable]
+	[/filter_conditional]
+	[set_vartiables]
```
Comment (ProditorMagnus) [inline data/campaigns/World_Conquest_II/MODDING.txt:25]: Reminder to check for typos if accepted

**Comment by Pentarctagon** [comment]
For the tropical terrain, I'd say to just compare the 1.10 version of the terrain used here with the current mainline terrain and see if it actually looks better or not.  By default though my assumption would be that it was kept more because of the subjective preference of the original author than because there was something worse about the terrain that replaced it.

**Comment by gfgtdf** [comment]
i removed all custom terrains for now for the following reasons:
- the village/forest terrain only exists to workaround #4894
- the ruined fence is not really mainline quality yet in my opinion
- i don't fell good about readding already deleted jungle terrain.

**Comment by Pentarctagon** [comment]
Make sure to check the travis jobs at some point as well - it looks like a preprocessor error is causing a bunch of the tests to fail.

**Comment by gfgtdf** [comment]
So the reason for the travis tests failing is 
```
Error (strict mode, strict_level = 1): wesnoth reported on channel warning general

20200521 00:26:59 warning general: no location found for 'data/campaigns/World_Conquest_II/translations', skipping textdomain
```
meaning it has no translations yet. I think this is not something i can fix before merging this.

**Comment by Pentarctagon** [comment]
The only place I see that mentioned is in the _main.cfg:
```
[textdomain]
	name="wesnoth-World_Conquest_II"
	path=data/campaigns/World_Conquest_II/translations
[/textdomain]
```
Whereas in Dead Water for example is has simply:
```
[textdomain]
    name="wesnoth-dw"
[/textdomain]
```
So is that `path` attribute really needed?

Additionally, the `name` attribute for other campaigns is `wesnoth-<lowercase acronym>` - so "wesnoth-dw" for Dead Water or "wesnoth-l" for Liberty - so the textdomain name here should be "wesnoth-wcii" for consistency I think.

**Comment by gfgtdf** [comment]
The scema validator complains quite a bit: first, it says 

```
20200521 00:25:39 error validation: Invalid key 'path=' in tag [textdomain]

at campaigns/World_Conquest_II/_main.cfg:7

    included from data/_main.cfg:38


```

Which i'm unsure about since the wiki says [textdomain] must use the path attribute=

 Also `except` seems to be just missing from the schema:
```
20200521 00:25:39 error validation: Invalid key 'except=' in tag [multiplayer_side]
at campaigns/World_Conquest_II/./utils/./era/./factions/Random.cfg:9
    included from campaigns/World_Conquest_II/./utils/./era/utilities.cfg:2
    included from campaigns/World_Conquest_II/./utils/utilities.cfg:2
    included from campaigns/World_Conquest_II/_main.cfg:29
    included from data/_main.cfg:38
    included from campaigns/World_Conquest_II/./utils/./era/utilities.cfg:12
    included from campaigns/World_Conquest_II/./utils/utilities.cfg:2
    included from campaigns/World_Conquest_II/_main.cfg:29
    included from data/_main.cfg:38
    included from campaigns/World_Conquest_II/_main.cfg:32
    included from data/_main.cfg:38

```
Furthermore it complains a about a few tags/atributes that are spcific to WCII:

```
20200521 00:25:39 error validation: Tag [scenario] may not be used in [generator]
at campaigns/World_Conquest_II/./scenarios/WC_II_scenario.cfg:71
    included from campaigns/World_Conquest_II/_main.cfg:30
    included from data/_main.cfg:38
    included from campaigns/World_Conquest_II/./scenarios/WC_II_scenario.cfg:87
    included from campaigns/World_Conquest_II/_main.cfg:30
    included from data/_main.cfg:38
20200521 00:25:39 error validation: Tag [scenario] may not be used in [generator]
at campaigns/World_Conquest_II/./scenarios/WC_II_scenario.cfg:71
    included from campaigns/World_Conquest_II/_main.cfg:30
    included from data/_main.cfg:38
    included from campaigns/World_Conquest_II/./scenarios/WC_II_scenario.cfg:88
    included from campaigns/World_Conquest_II/_main.cfg:30
    included from data/_main.cfg:38
20200521 00:25:39 error validation: Tag [scenario] may not be used in [generator]
at campaigns/World_Conquest_II/./scenarios/WC_II_scenario.cfg:71
    included from campaigns/World_Conquest_II/_main.cfg:30
    included from data/_main.cfg:38
    included from campaigns/World_Conquest_II/./scenarios/WC_II_scenario.cfg:89
    included from campaigns/World_Conquest_II/_main.cfg:30
    included from data/_main.cfg:38
20200521 00:25:39 error validation: Tag [wc2_extra] may not be used in [multiplayer_side]
at campaigns/World_Conquest_II/./utils/./era/./factions/The_Trust.cfg:11
    included from campaigns/World_Conquest_II/./utils/./era/utilities.cfg:2
    included from campaigns/World_Conquest_II/./utils/utilities.cfg:2
    included from campaigns/World_Conquest_II/_main.cfg:29
    included from data/_main.cfg:38
    included from campaigns/World_Conquest_II/./utils/./era/utilities.cfg:13
    included from campaigns/World_Conquest_II/./utils/utilities.cfg:2
    included from campaigns/World_Conquest_II/_main.cfg:29
    included from data/_main.cfg:38
    included from campaigns/World_Conquest_II/_main.cfg:32
    included from data/_main.cfg:38
20200521 00:25:39 error validation: Tag [wc2_extra] may not be used in [multiplayer_side]
at campaigns/World_Conquest_II/./utils/./era/./factions/The_Guild.cfg:11
    included from campaigns/World_Conquest_II/./utils/./era/utilities.cfg:2
    included from campaigns/World_Conquest_II/./utils/utilities.cfg:2
    included from campaigns/World_Conquest_II/_main.cfg:29
    included from data/_main.cfg:38
    included from campaigns/World_Conquest_II/./utils/./era/utilities.cfg:14
    included from campaigns/World_Conquest_II/./utils/utilities.cfg:2
    included from campaigns/World_Conquest_II/_main.cfg:29
    included from data/_main.cfg:38
    included from campaigns/World_Conquest_II/_main.cfg:32
    included from data/_main.cfg:38

20200521 00:25:39 error validation: Tag [wc2_extra] may not be used in [era]
at campaigns/World_Conquest_II/./utils/./era/utilities.cfg:25
    included from campaigns/World_Conquest_II/./utils/utilities.cfg:2
    included from campaigns/World_Conquest_II/_main.cfg:29
    included from data/_main.cfg:38
    included from campaigns/World_Conquest_II/_main.cfg:32
    included from data/_main.cfg:38
```

**Comment by Pentarctagon** [comment]
The schema validator not knowing all valid tags/attributes isn't a new issue to have happen - see 6dd916283ccdfc2842fafbe5450f09f18105bcf8 from #4882 for a recent example.  So the `path` and `except` attributes should be added to it (though it looks like mainline campaigns don't use `path`).

For `wc2_extra`, it should also be added to the schema validator.  Though also I'd say it should be moved to the same place as the rest of the core lua scripts that define tags meant for UMC authors, renamed to something more descriptive like [enable_world_conquest], and everything related to its use documented on the wiki once its been merged rather than in MODDING.txt.

Also, unrelatedly, the _info.cfg file should be deleted.

**Comment by gfgtdf** [comment]
> For wc2_extra, it should also be added to the schema validator. Though also I'd say it should be moved to the same place as the rest of the core lua scripts that define tags meant for UMC authors,

Well its not an actionwml tag just a data tag, so that technicially not possible.

>  renamed to something more descriptive like [enable_world_conquest],

that would be a misleading name since the presence of these doesn't enable world conquest II if world conquerst II is not present these tags have no effect.

**Comment by Pentarctagon** [comment]
> Well its not an actionwml tag just a data tag, so that technicially not possible.

Ah, alright, my bad.

> that would be a misleading name since the presence of these doesn't enable world conquest II if world conquerst II is not present these tags have no effect.

Maybe [world_conquest_data] then?  IMO having an acronym in a tag name isn't great, and "extra" is just really vague.

World Conquest would also always be available if it's mainlined, wouldn't it?

**Comment by gfgtdf** [comment]
> World Conquest would also always be available if it's mainlined, wouldn't it?

How exactly do you mean this?

**Comment by soliton-** [comment]
As an aside, I suggest to just call it world conquest in mainline. I think it would be less confusing in the long run.

**Comment by Pentarctagon** [comment]
> How exactly do you mean this?

I mean that if it's mainlined then by definition it will be included in every Wesnoth installation from that point onward, just like how every Wesnoth installation has Heir to the Throne, Isar's Cross, etc (excluding linux distros that package all the campaigns individually, since that's never been supported anyway).  In which case, there would never be an instance where WCII isn't present.

**Comment by gfgtdf** [comment]
Well, installed but not 'active', just like any.otger events, the event/lua tags that run wc2 will no be present unless you include them in your scenario, you can however add them to your scenario by adding ` [load_ressource] id=wc2_scenario_res` in your scenario.

**Comment by gfgtdf** [comment]
> As an aside, I suggest to just call it world conquest in mainline. I think it would be less confusing in the long run.

I don't have a strong opinion on this one.

**Comment by gfgtdf** [comment]
so the travis issues are fixed now, i personally see nothing that needs to be done before merging.

Some things are left todo, like deciding whether we want to remove the " II" from the name for example  but that could also be done afterwards

**Comment by Pentarctagon** [comment]
The name and the [wc2_extra] tag name are the two remaining things, I think?

**Comment by gfgtdf** [comment]
> The name

So is there a a consent now that it should be renamed? So far i only had solitons opinion and i didnt want to do it unless i know that the majority agrees.

>  [wc2_extra] tag name

Hmm yes, we should rename that one aswell then, also currently all internal wml variables that are used by this addon are prefixed by `wc2_` i could replace them by for example `woco_` not sure about that.

**Comment by Pentarctagon** [comment]
> So is there a a consent now that it should be renamed? So far i only had solitons opinion and i didnt want to do it unless i know that the majority agrees.

I'll leave it up to @soliton- if he feels strongly about it.  I'm kind of ambivalent about the name change, since it would also potentially be a bit confusing for existing players who already know it as World Conquest II.  I don't think it'll be hard for people to adjust either way though.

> Hmm yes, we should rename that one aswell then, also currently all internal wml variables that are used by this addon are prefixed by wc2_ i could replace them by for example woco_ not sure about that. (this would basicailyl cahnge the entire code ina way that woudl make it hard for me to backport changes to 1.14, but if its wanted i can do this.)

Internal variable names I'm less concerned about honestly, since they're, well, internal.  Maybe leave an open issue to rename them just before 1.16.0 is released?

The tag name though I'd definitely like to be more descriptive, like [world_conquest_data], since it's something UMC authors would be using.

**Comment by soliton-** [comment]
> So is there a a consent now that it should be renamed? So far i only had solitons opinion and i didnt want to do it unless i know that the majority agrees.

I don't think you'll get a majority opinion on here. Perhaps do a forum poll if you really want to know.

As already said I think in the long run it will be less confusing to omit the II because there is no original world conquest for 1.14 and world conquest II is more like a successor than a different variant that requires a new name anyway.

Another thing is addons based on world conquest where stuff like the already mentioned `[wc2_extra]` tag comes in and is just less straight forward and raises questions for anyone new if we keep the 2.

Anyhow I'm not trying to veto this without the rename or anything. If you want to keep it as close as possible to the 1.14 version for ease of maintenance I can understand that.

**Comment by gfgtdf** [comment]
> Anyhow I'm not trying to veto this without the rename or anything. If you want to keep it as close as possible to the 1.14 version for ease of maintenance I can understand that.


yes i want, but this doesn't apply to the wc2_extra tag or the name. just for the name of the lua functions.

**Comment by gfgtdf** [comment]
So i changed the name of the campaign and the tagname  now.

**Comment by Pentarctagon** [comment]
I'd say the last thing then would just be to double check the text for typos.  For example, in STRUCTURE.txt:
` A describtion of this addons code: `
should be
` A description of this addon's code: `

**Comment by gfgtdf** [comment]
So im my side in quite happy with the current state and would probably merge it as-is

**Comment by Pentarctagon** [comment]
MP and Editor WML validation is failing: https://travis-ci.com/github/wesnoth/wesnoth/jobs/349471987#L1068-L1135

**Comment by Pentarctagon** [comment]
(I added myself for review/approval to get a screenshot of it, nothing else)

File: data/campaigns/World_Conquest/scenarios/WC_II_scenario.cfg [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+#define WC2_CAMPAIGN_NEW PLAYERS
+## Everything that usually goes into [scenario] like [event], [load_ressource], map_data 
+## etc is generated by lua. This is also why there is only one [multiplayer] tag: the next
+## scenario is again "WC_II_{PLAYERS}p" and the lua code detects that is was called a second
+## time and generated the second scenario. (same for third, fourth and fifth). All the lua
+## code inside ./lua/map does map/scenario generation.
+[multiplayer]
+	id = "WC_II_{PLAYERS}p{ID_SUFFIX}"
+	name= {WC_II_CAMPAIGN_NAME_{PLAYERS}P} + {ID_SUFFIX}
```
Comment (Adriankhl) [inline data/campaigns/World_Conquest/scenarios/WC_II_scenario.cfg:33]: The description is not correctly display. 
From what I have tested, it should be `name= {WC_II_CAMPAIGN_NAME_{PLAYERS}P} + "{ID_SUFFIX}"`.

**Comment by Adriankhl** [comment]
And I believe that the commit messages should be improved (such as the `f schema` message). Maybe follow the message style from other campaigns - use `WC: ` to highlight World Conquest specific commits, and refactor the commits a little bit.

**Comment by gfgtdf** [comment]
> And I believe that the commit messages should be improved

some of the commits, in particular the "f schema" will be merged into other commits before merging. I just didn't want to rebase everytime i make a new commit.

File: data/campaigns/World_Conquest/scenarios/WC_II_scenario.cfg (hunk 2 above)
Comment (gfgtdf) [reply data/campaigns/World_Conquest/scenarios/WC_II_scenario.cfg:33]: oh yes probably, i just tested it with a  nonempty ID_SUFFIX yet.

**Pull request merged**